### DISCUSSION TITLE
Question on Jointjs+ Navigator

### DISCUSSION THREAD

**Discussion by ROYopedia** [clientIO/joint#2424] [Q&A]
### What happened?

Hello Team,
I am using joinjts+ navigator in our project.
Now I have requirement where I am hiding elements on paper based on some user event.

```
export function setCellVisibility(cell: dia.Cell, visible: boolean): void {
  cell.set("hidden", !visible);
}
```

Now, Wewn user is hiding cells on paper the navigator is unable to hide it from its internal viewer.
in Navigator viewer the cells are not hidden.
Any idea?

### Version

3.7

### What browsers are you seeing the problem on?

Chrome

### What operating system are you seeing the problem on?

Linux

**Comment by kumilingus** [comment]
How do you hide the cells in the main paper? You have to do the same with the `Navigator` paper.

**Reply by ROYopedia** [reply]
By setting cell.set("hidden",false/true)

Below is one code example

``` 
export function setCellVisibility(cell: dia.Cell, visible: boolean): void {
  cell.set("hidden", !visible);
}

```
 ```
 paper.on("node:collapse", (elementView: dia.ElementView, evt) => {
    evt.stopPropagation();
    const node = elementView.model as Node;

      setCellVisibility(node, false);
    });
```


  });

**Reply by kumilingus** [reply]
This code does not hide the DOM element. This code just sets the model property called `hidden`.

**Comment by ROYopedia** [comment]
It is hiding the cells on paper indeed. I took reference from one of the demo called DWDM circuit.
https://www.jointjs.com/demos/dwdm-circuit

And the github link to the file 
https://github.com/clientIO/joint/blob/master/demo/dwdm/src/shapes/utils.ts.
Example:
In demo if we add a navigator and collapse via - icon. In paper the cell along with its child elements getiing hidden.
But in navigator it is still available as it is.

**Reply by kumilingus** [reply]
The elements in the DWDM demo are hidden using the paper's [viewport option](https://github.com/clientIO/joint/blob/master/demo/dwdm/src/index.ts#L38).

You have to add the `viewport` option to the `Navigator` as well. For that use Navigator's `paperOptions` option.

**Reply by ROYopedia** [reply]
Can you give an example with respect to the demo?

I tried this but seems not working/hiding
```
  nav = new ui.Navigator({
      paperScroller: paperScroller,
      width: 300,
      height: 100,
      padding: 2,
      zoomOptions: { max: 3, min: 0.2 },
      viewport: (view: dia.CellView) => isCellHidden(view.model)
    });

```

Do I need to update or re render the navigator?

**Reply by kumilingus** [reply]
```js
nav = new ui.Navigator({
    paperScroller: paperScroller,
    width: 300,
    height: 100,
    padding: 2,
    zoomOptions: { max: 3, min: 0.2 },
    paperOptions: {
      viewport: (view: dia.CellView) => isCellHidden(view.model)
    }
  });
 ```

**Reply by ROYopedia** [reply]
It is not working as expected. I tried in many way using viewport option. 
But it is not working. 
Will it be possible for you to add a navigator to the demo called DWDM circuit.
https://www.jointjs.com/demos/dwdm-circuit 
And check if it has any bug?

**Reply by kumilingus** [reply]
Try to make the Navigator asynchronous. In synchronous mode, you must use the [checkViewport()](https://resources.jointjs.com/docs/jointjs/v3.7/joint.html#dia.Paper.prototype.checkViewport) method.
```js
nav = new ui.Navigator({
    paperScroller: paperScroller,
    width: 300,
    height: 100,
    padding: 2,
    zoomOptions: { max: 3, min: 0.2 },
    paperOptions: {
      async: true,
      sorting: dia.Paper.sorting.APPROX,
      viewport: (view: dia.CellView) => isCellHidden(view.model)
    }
  });
```

**Reply by ROYopedia** [reply]
Thanks for the update.

Currently when I am collapsing elements  in paper the navigator element is also collapsing after the async:true option.
But the links are scattered acrros the navigator. While in actual paper it is repositioning and pointing correctly as required.

Is there any option available for link reparent or update option when we r collapsing for navigator?

**Reply by kumilingus** [reply]
What exactly do you do with the links in the main paper? You need to do the same thing with the paper inside the navigator.

_Note: There is no difference between the main paper and the paper inside the navigator._

**Reply by ROYopedia** [reply]
Referring to the same example given in joinjtjs DWDM circuit.
https://www.jointjs.com/demos/dwdm-circuit

Here if you see while collpasing(Node level collapsing into minimize node version with background image) the links are re-positioning accordingly.t:

But in same demo if you add a navigator and when you collapse the node in navigator view the link is  connection is scattered.

example screenshot
![image](https://github.com/clientIO/joint/assets/22453886/68b180e0-b2b0-45b9-8b03-f491133e0690)

I collapse the first node and If you see the link is not proper which is connecting to node 2 for navigator but working fine in main paper.If the paper is same then it should also behave the same right?


Any idea?

**Reply by kumilingus** [reply]
If you configure it in the same way then yes.

Try to add these options to the navigator paper: https://github.com/clientIO/joint/blob/master/demo/dwdm/src/index.ts#L25-L37

**Reply by ROYopedia** [reply]
All the options are same as main paper. Still no luck.
Same scattered links. It should work same as the both paper are same. 
Can you please give it a try? or any other way?

**Reply by kumilingus** [reply]
Please send the _DWDM_ application enhanced with the Navigator to `support@jointjs.com` and we will take a look and let you know.

**Reply by ROYopedia** [reply]
Sure. I will have a internal discussion and send accordingly.

**Reply by kumilingus** [reply]
Just to clarify, there is no need to send us your actual application. You can take the DWDM example, replace JJ with JJ+, add the Navigator there, and send it to us (helping us save time as this is a free support channel). Thank you.
